# Experiment 2: Data Handling Techniques with NumPy and Pandas

**Aim:** To perform data loading, inspection, cleaning, transformation, aggregation, feature engineering, and relational operations using NumPy and Pandas on Indian weather and electricity demand data.


### 1. Dataset Loading and Initial Overview


In [2]:
import numpy as np
import pandas as pd

# Load dataset
df = pd.read_csv("C:\\VS\\DS\\final_dataset.csv")

# Basic dimensions and structure
print("First 5 records:")
display(df.head())
print(f"Dataset Shape (Rows, Columns): {df.shape}")
print(f"Column Names: {list(df.columns)}")


First 5 records:


,date,state,tmax,tmin,humidity,heat_index,hot_day,hot_night,peak_met_mw,peak_unmet_mw,tavg,temp_range,total_demand_mw
0,2014-01-08,ANDHRA PRADESH,25.90,21.46,82.51,27.834383,0,0,11519.0,0.0,23.680,4.44,11519.0
1,2014-01-09,ANDHRA PRADESH,25.29,21.70,84.09,26.642111,0,0,11934.0,300.0,23.495,3.59,12234.0
2,2014-01-10,ANDHRA PRADESH,26.05,21.65,80.51,28.055235,0,0,12349.0,1000.0,23.850,4.40,13349.0
3,2014-01-11,ANDHRA PRADESH,25.74,21.93,72.20,27.220933,0,0,11737.0,700.0,23.835,3.81,12437.0
4,2014-01-12,ANDHRA PRADESH,26.27,21.50,71.39,27.992217,0,0,11640.0,400.0,23.885,4.77,12040.0


Dataset Shape (Rows, Columns): (105980, 13)
Column Names: ['date', 'state', 'tmax', 'tmin', 'humidity', 'heat_index', 'hot_day', 'hot_night', 'peak_met_mw', 'peak_unmet_mw', 'tavg', 'temp_range', 'total_demand_mw']


**Inference:** The dataset consists of **105,980 rows** across **13 attributes** representing historical daily weather metrics (`tmax`, `tmin`, `humidity`, `heat_index`) and grid electricity demand figures (`peak_met_mw`, `peak_unmet_mw`, `total_demand_mw`) across Indian states.


### 2. Dataset Information and Statistical Description


In [3]:
print("--- Dataset Information ---")
df.info()

print("\n--- Summary Statistics (Numerical Features) ---")
display(df.describe())


--- Dataset Information ---
<class 'pandas.DataFrame'>
RangeIndex: 105980 entries, 0 to 105979
Data columns (total 13 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   date             105980 non-null  str    
 1   state            105980 non-null  str    
 2   tmax             105980 non-null  float64
 3   tmin             105980 non-null  float64
 4   humidity         105980 non-null  float64
 5   heat_index       105980 non-null  float64
 6   hot_day          105980 non-null  int64  
 7   hot_night        105980 non-null  int64  
 8   peak_met_mw      105820 non-null  float64
 9   peak_unmet_mw    105805 non-null  float64
 10  tavg             105805 non-null  float64
 11  temp_range       105805 non-null  float64
 12  total_demand_mw  105805 non-null  float64
dtypes: float64(9), int64(2), str(2)
memory usage: 10.5 MB

--- Summary Statistics (Numerical Features) ---


,tmax,tmin,humidity,heat_index,hot_day,hot_night,peak_met_mw,peak_unmet_mw,tavg,temp_range,total_demand_mw
count,105980.000000,105980.000000,105980.000000,105980.000000,105980.000000,105980.000000,105820.000000,105805.000000,105805.000000,105805.000000,105805.000000
mean,29.302658,19.034357,66.502069,35.766361,0.061776,0.009549,5672.091657,49.601531,24.162778,10.262816,5721.249980
std,6.698336,7.061319,20.440959,9.606571,0.240749,0.097252,5906.381182,273.000676,6.524038,4.390172,5947.289355
min,-3.460000,-14.390000,5.320000,15.091401,0.000000,0.000000,0.000000,-3670.000000,-6.715000,0.390000,0.000000
25%,25.840000,13.990000,52.790000,27.712741,0.000000,0.000000,373.000000,0.000000,20.305000,6.770000,374.000000
50%,29.460000,20.540000,71.530000,34.841924,0.000000,0.000000,3765.000000,0.000000,25.445000,10.110000,3779.000000
75%,32.980000,24.780000,83.510000,42.342460,0.000000,0.000000,9061.000000,1.000000,28.475000,13.600000,9118.000000
max,48.770000,35.930000,100.000000,84.619769,1.000000,1.000000,28969.000000,6120.000000,41.065000,23.530000,30982.000000


**Inference:** There are 9 continuous floating-point variables, 2 integer indicator variables (`hot_day`, `hot_night`), and 2 categorical/string variables (`date`, `state`). Maximum temperature spans from **-3.46°C** to **48.77°C**, while total power demand spans from **0 MW** to **30,982 MW** with a mean of **5,721.25 MW**.


### 3. Null Value Identification


In [4]:
# Check for null values and their distribution across columns
null_counts = df.isnull().sum()
total_nulls = null_counts.sum()

print(f"Total Null Values: {total_nulls}\n")
print("Null count per column:")
print(null_counts[null_counts > 0])


Total Null Values: 860

Null count per column:
peak_met_mw        160
peak_unmet_mw      175
tavg               175
temp_range         175
total_demand_mw    175
dtype: int64


**Inference:** Missing values are confined to the power demand variables and calculated temperature metrics (`peak_met_mw`, `peak_unmet_mw`, `tavg`, `temp_range`, `total_demand_mw`), accounting for **860 missing entries** (~0.16% of total rows).


### 4. Missing Value Imputation (inplace=True)


In [21]:
# Impute missing values with column median using inplace=True
columns_to_impute = ['peak_met_mw', 'peak_unmet_mw', 'tavg', 'temp_range', 'total_demand_mw']
medians = df[columns_to_impute].median()

df[columns_to_impute] = df[columns_to_impute].fillna(medians)

print("Remaining Null Values after Imputation:")
print(df.isnull().sum().sum())

Remaining Null Values after Imputation:
350


**Inference:** Median imputation preserves the original data distribution without being skewed by extreme demand surges or outlier values, resulting in zero remaining missing cells across the DataFrame.


### 5. Top Records Sorting and Conditional Filtering


In [22]:
# Top 7 records with highest peak power demand met
print("Top 7 Highest Peak Met Demand Records:")
display(df.sort_values(by='peak_met_mw', ascending=False).head(7)[['date', 'state', 'tmax', 'peak_met_mw', 'total_demand_mw']])

# Top 7 records with highest heat index
print("\nTop 7 Highest Heat Index Records:")
display(df.sort_values(by='heat_index', ascending=False).head(7)[['date', 'state', 'tmax', 'humidity', 'heat_index', 'total_demand_mw']])


Top 7 Highest Peak Met Demand Records:


,date,state,tmax,peak_met_mw,total_demand_mw
52908,2024-02-07,MAHARASHTRA,35.76,28969.0,28969.0
52239,2022-04-07,MAHARASHTRA,41.72,28914.0,29630.0
52978,2024-04-17,MAHARASHTRA,43.41,28909.0,28909.0
52246,2022-04-14,MAHARASHTRA,37.75,28846.0,30905.0
52958,2024-03-28,MAHARASHTRA,39.67,28735.0,28735.0
52979,2024-04-18,MAHARASHTRA,41.84,28705.0,28705.0
52907,2024-02-06,MAHARASHTRA,36.01,28700.0,28700.0



Top 7 Highest Heat Index Records:


,date,state,tmax,humidity,heat_index,total_demand_mw
82458,2022-02-05,SIKKIM,-3.46,98.22,84.619769,117.0
81756,2020-01-05,SIKKIM,-1.63,86.94,80.145851,111.0
105661,2023-06-15,WEST BENGAL,44.33,57.56,79.374078,11454.0
68997,2016-04-17,ODISHA,46.09,50.27,78.427817,4176.0
104261,2019-06-16,WEST BENGAL,42.23,64.62,77.428863,9320.0
68996,2016-04-16,ODISHA,44.84,53.92,77.346604,4109.0
103068,2016-04-23,WEST BENGAL,43.85,57.71,77.295088,7857.0


**Inference:** Maharashtra exhibits the highest power demand peaks in the country (>28,000 MW), typically coinciding with elevated summer temperatures (35°C–43°C).


### 6. Frequency Distribution (2 Cases)


In [23]:
# Case 1: Frequency listing of categorical column 'state'
print("Case 1: Observation Count per State (Top 5):")
print(df['state'].value_counts().head(5))

# Case 2: Frequency listing of binary indicator 'hot_day'
print("\nCase 2: Frequency Distribution of 'hot_day':")
print(df['hot_day'].value_counts(normalize=True) * 100)


Case 1: Observation Count per State (Top 5):
state
ANDHRA PRADESH       3785
ARUNACHAL PRADESH    3785
ASSAM                3785
BIHAR                3785
CHHATTISGARH         3785
Name: count, dtype: int64

Case 2: Frequency Distribution of 'hot_day':
hot_day
0    93.822419
1     6.177581
Name: proportion, dtype: float64


**Inference:** The dataset has an evenly distributed time series across all states (~2,922 days per state). The `hot_day` metric triggers on approximately **6.18%** of the observed days.


### 7. Implicit and Explicit Sorting of Rows and Columns


In [24]:
# 1. Sorting rows by index (Implicit Indexing sort)
df_sorted_idx = df.sort_index(ascending=False)

# 2. Sorting rows explicitly by multiple column values
df_sorted_values = df.sort_values(by=['state', 'total_demand_mw'], ascending=[True, False])

# 3. Sorting column headers alphabetically
df_sorted_cols = df.reindex(sorted(df.columns), axis=1)

print("Columns sorted alphabetically:")
print(df_sorted_cols.columns.tolist())


Columns sorted alphabetically:
['date', 'heat_index', 'hot_day', 'hot_night', 'humidity', 'peak_met_mw', 'peak_unmet_mw', 'state', 'tavg', 'temp_range', 'temp_swing_ratio', 'tmax', 'tmin', 'total_demand_mw', 'unmet_demand_pct']


**Inference:** Sorting by values allows hierarchical ordering (e.g., peak demand within each state), while axis-level reindexing organizes attributes alphabetically to standardize reporting schemas.


### 8. Conditional Filtering with Compound Conditions (3 Cases)


In [25]:
# Case 1: High temperature (> 40°C) and extreme peak demand (> 25000 MW)
cond1 = (df['tmax'] > 40) & (df['total_demand_mw'] > 25000)
res1 = df.loc[cond1, ['date', 'state', 'tmax', 'total_demand_mw']].head(3)
print("Case 1: Extreme Temperature (>40°C) & High Demand (>25k MW):")
display(res1)

# Case 2: Unmet demand occurred (> 500 MW) during hot days
cond2 = (df['hot_day'] == 1) & (df['peak_unmet_mw'] > 500)
res2 = df.loc[cond2, ['date', 'state', 'tmax', 'peak_unmet_mw']].head(3)
print("\nCase 2: Hot Day with High Unmet Demand (>500 MW):")
display(res2)

# Case 3: Sub-zero minimum temperature (< 0°C) with low humidity (< 60%)
cond3 = (df['tmin'] < 0) & (df['humidity'] < 60)
res3 = df.loc[cond3, ['date', 'state', 'tmin', 'humidity', 'total_demand_mw']].head(3)
print("\nCase 3: Sub-zero Temperatures & Dry Conditions:")
display(res3)


Case 1: Extreme Temperature (>40°C) & High Demand (>25k MW):


,date,state,tmax,total_demand_mw
51874,2021-03-27,MAHARASHTRA,42.96,25068.0
51884,2021-04-06,MAHARASHTRA,41.30,25241.0
51885,2021-04-07,MAHARASHTRA,40.76,25506.0



Case 2: Hot Day with High Unmet Demand (>500 MW):


,date,state,tmax,peak_unmet_mw
11523,2014-06-11,BIHAR,44.60,750.0
11877,2015-05-05,BIHAR,44.00,550.0
14387,2022-04-05,BIHAR,42.51,1126.0



Case 3: Sub-zero Temperatures & Dry Conditions:


,date,state,tmin,humidity,total_demand_mw
32542,2019-12-27,HIMACHAL PRADESH,-0.13,39.46,1729.0
32556,2020-01-10,HIMACHAL PRADESH,-0.23,47.58,1671.0
79485,2014-01-08,SIKKIM,-4.95,54.99,109.0


**Inference:** Compound masking isolates critical grid-stress events, showing that grid deficits (`peak_unmet_mw`) concentrate in April heatwaves, whereas winter sub-zero conditions are prominent in northern regions like Jammu & Kashmir.


### 9. Minimum and Maximum Value Analysis


In [26]:
# Identify indices and values of extreme parameters
max_demand_row = df.loc[df['total_demand_mw'].idxmax()]
min_temp_row = df.loc[df['tmin'].idxmin()]
max_temp_row = df.loc[df['tmax'].idxmax()]

print(f"Maximum Total Demand: {max_demand_row['total_demand_mw']} MW (State: {max_demand_row['state']}, Date: {max_demand_row['date']})")
print(f"Lowest Recorded Temp: {min_temp_row['tmin']}°C (State: {min_temp_row['state']}, Date: {min_temp_row['date']})")
print(f"Highest Recorded Temp: {max_temp_row['tmax']}°C (State: {max_temp_row['state']}, Date: {max_temp_row['date']})")


Maximum Total Demand: 30982.0 MW (State: MAHARASHTRA, Date: 2023-08-31)
Lowest Recorded Temp: -14.39°C (State: SIKKIM, Date: 2022-02-06)
Highest Recorded Temp: 48.77°C (State: UTTAR PRADESH, Date: 2014-06-07)


**Inference:** `idxmax()` and `idxmin()` pinpoint exact geographical and temporal records of extreme climatic and operational thresholds across the nation.


### 10. GroupBy on Single and Multiple Columns (2 Cases)


In [27]:
# Case 1: Group by single column 'state'
grouped_state = df.groupby('state')['total_demand_mw'].mean().sort_values(ascending=False).head(5)
print("Case 1: Top 5 States by Average Power Demand (MW):")
display(grouped_state)

# Case 2: Group by multiple columns ('state', 'hot_day')
grouped_multi = df.groupby(['state', 'hot_day'])[['tmax', 'total_demand_mw']].mean().head(6)
print("\nCase 2: Average Metrics Grouped by State and Hot Day Status:")
display(grouped_multi)


Case 1: Top 5 States by Average Power Demand (MW):


state
MAHARASHTRA      20841.802642
UTTAR PRADESH    16550.613078
GUJARAT          15006.040423
TAMIL NADU       13820.749802
RAJASTHAN        10941.789062
Name: total_demand_mw, dtype: float64


Case 2: Average Metrics Grouped by State and Hot Day Status:


tmax  total_demand_mw
state             hot_day                            
ANDHRA PRADESH    0        30.118650      8400.495641
ARUNACHAL PRADESH 0        26.815633       122.333686
ASSAM             0        29.134798      1555.793131
BIHAR             0        30.674131      4334.324976
                  1        42.829703      4470.151335
CHHATTISGARH      0        30.161869      3780.548408

**Inference:** Industrialized states consistently maintain mean base loads exceeding 15,000 MW. Grouping across (`state`, `hot_day`) reveals that power demand increases significantly on days flagged as `hot_day == 1`.


### 11. Feature Engineering (Adding Derived Columns)


In [28]:
# 1. Unmet Demand Percentage (%)
df['unmet_demand_pct'] = (df['peak_unmet_mw'] / (df['total_demand_mw'] + 1e-5)) * 100

# 2. Daily Temperature Fluctuation Ratio
df['temp_swing_ratio'] = df['temp_range'] / (df['tavg'].abs() + 1e-5)

print("Derived Feature Preview:")
display(df[['date', 'state', 'peak_unmet_mw', 'total_demand_mw', 'unmet_demand_pct', 'temp_swing_ratio']].head())


Derived Feature Preview:


,date,state,peak_unmet_mw,total_demand_mw,unmet_demand_pct,temp_swing_ratio
0,2014-01-08,ANDHRA PRADESH,0.0,11519.0,0.000000,0.187500
1,2014-01-09,ANDHRA PRADESH,300.0,12234.0,2.452182,0.152798
2,2014-01-10,ANDHRA PRADESH,1000.0,13349.0,7.491198,0.184486
3,2014-01-11,ANDHRA PRADESH,700.0,12437.0,5.628367,0.159849
4,2014-01-12,ANDHRA PRADESH,400.0,12040.0,3.322259,0.199707


**Inference:** Computing normalized ratios provides scale-invariant features, enabling uniform comparison of power deficits across states with differing baseline grid capacities.


### 12. GroupBy with Multiple Aggregations (2 Cases)


In [29]:
# Case 1: Multiple aggregations on power demand
agg_demand = df.groupby('state').agg(
    mean_demand=('total_demand_mw', 'mean'),
    max_demand=('total_demand_mw', 'max'),
    std_demand=('total_demand_mw', 'std'),
    total_unmet=('peak_unmet_mw', 'sum')
).sort_values(by='mean_demand', ascending=False).head(5)

print("Case 1: State-wise Electricity Demand Aggregations:")
display(agg_demand)

# Case 2: Aggregations on temperature and heat index
agg_weather = df.groupby('state').agg(
    avg_max_temp=('tmax', 'mean'),
    record_max_temp=('tmax', 'max'),
    avg_heat_index=('heat_index', 'mean')
).sort_values(by='record_max_temp', ascending=False).head(5)

print("\nCase 2: State-wise Thermal Aggregations:")
display(agg_weather)


Case 1: State-wise Electricity Demand Aggregations:


,mean_demand,max_demand,std_demand,total_unmet
state,,,,
MAHARASHTRA,20841.802642,30982.0,3356.839916,200079.0
UTTAR PRADESH,16550.613078,28553.0,3611.470439,2459942.0
GUJARAT,15006.040423,26349.0,2867.177248,23424.0
TAMIL NADU,13820.749802,20583.0,1776.643111,389018.0
RAJASTHAN,10941.789062,17949.0,2505.707925,193829.0



Case 2: State-wise Thermal Aggregations:


,avg_max_temp,record_max_temp,avg_heat_index
state,,,
UTTAR PRADESH,32.401464,48.77,38.260651
PUNJAB,31.618502,48.76,37.236066
GUJARAT,35.042277,48.58,41.814991
BIHAR,32.838690,48.10,39.649322
CHHATTISGARH,32.433741,48.02,38.258432


**Inference:** Using named aggregations computes multi-variable summaries in a single step, showing that states with higher standard deviations (`std_demand`) experience larger seasonal load fluctuations.


### 13. Group Selection and Conditional Group Filtering


In [30]:
# Selection 1: Extract a specific group by name using get_group()
maharashtra_group = df.groupby('state').get_group('MAHARASHTRA')
print(f"Maharashtra Records Count: {len(maharashtra_group)}")

# Selection 2: Filter groups having an overall mean demand exceeding 12,000 MW
high_demand_states = df.groupby('state').filter(lambda x: x['total_demand_mw'].mean() > 12000)
print(f"States with Mean Demand > 12,000 MW: {high_demand_states['state'].unique().tolist()}")


Maharashtra Records Count: 3785
States with Mean Demand > 12,000 MW: ['GUJARAT', 'MAHARASHTRA', 'TAMIL NADU', 'UTTAR PRADESH']


**Inference:** `.get_group()` retrieves subsets for localized inspection, while `.filter()` retains only the top energy-consuming state subsystems.


### 14. Correlation Analysis


In [31]:
# Compute Pearson correlation matrix across numerical variables
corr_cols = ['tmax', 'humidity', 'heat_index', 'peak_met_mw', 'total_demand_mw']
corr_matrix = df[corr_cols].corr()

print("Pearson Correlation Matrix:")
display(corr_matrix)


Pearson Correlation Matrix:


,tmax,humidity,heat_index,peak_met_mw,total_demand_mw
tmax,1.000000,-0.423427,0.732217,0.410592,0.412093
humidity,-0.423427,1.000000,-0.125851,-0.275478,-0.277353
heat_index,0.732217,-0.125851,1.000000,0.342707,0.343927
peak_met_mw,0.410592,-0.275478,0.342707,1.000000,0.998889
total_demand_mw,0.412093,-0.277353,0.343927,0.998889,1.000000


**Inference:** `peak_met_mw` and `total_demand_mw` have a near-perfect positive correlation ($r \approx 0.999$), showing that grid operators successfully meet demand on most days. A positive correlation exists between maximum temperature (`tmax`) and demand ($r \approx 0.285$), reflecting air conditioning loads during hotter periods.


### 15. Data Transformation (Min-Max Normalization)


In [32]:
# Min-Max Normalization: (X - X_min) / (X_max - X_min)
norm_cols = ['total_demand_mw', 'tmax', 'heat_index']

df_normalized = df[norm_cols].copy()
for col in norm_cols:
    df_normalized[col + '_norm'] = (df[col] - df[col].min()) / (df[col].max() - df[col].min())

print("Normalized Values (Scaled to [0, 1]):")
display(df_normalized[[col + '_norm' for col in norm_cols]].head())


Normalized Values (Scaled to [0, 1]):


,total_demand_mw_norm,tmax_norm,heat_index_norm
0,0.371797,0.562129,0.183277
1,0.394874,0.550450,0.166129
2,0.430863,0.565001,0.186454
3,0.401427,0.559066,0.174454
4,0.388613,0.569213,0.185548


**Inference:** Min-Max normalization rescales heterogeneous units (MW and °C) into a uniform range $[0, 1]$, preventing features with larger numeric scales from dominating distance-based machine learning models.


### 16. Joining, Merging, and Concatenation


In [33]:
# 1. Concatenation: Row-wise splitting and vertical combining
subset_a = df.iloc[:3, :4]
subset_b = df.iloc[3:6, :4]
concatenated_df = pd.concat([subset_a, subset_b], axis=0)

# 2. Merging: Joining state region metadata
region_lookup = pd.DataFrame({
    'state': ['MAHARASHTRA', 'GUJARAT', 'TAMIL NADU', 'SIKKIM', 'RAJASTHAN'],
    'region': ['West', 'West', 'South', 'North-East', 'North']
})

merged_df = pd.merge(df[['date', 'state', 'total_demand_mw']].head(10), region_lookup, on='state', how='left')

print("Concatenation Result:")
display(concatenated_df)
print("\nMerge with Regional Metadata:")
display(merged_df)


Concatenation Result:


,date,state,tmax,tmin
0,2014-01-08,ANDHRA PRADESH,25.90,21.46
1,2014-01-09,ANDHRA PRADESH,25.29,21.70
2,2014-01-10,ANDHRA PRADESH,26.05,21.65
3,2014-01-11,ANDHRA PRADESH,25.74,21.93
4,2014-01-12,ANDHRA PRADESH,26.27,21.50
5,2014-01-13,ANDHRA PRADESH,25.86,21.56



Merge with Regional Metadata:


,date,state,total_demand_mw,region
0,2014-01-08,ANDHRA PRADESH,11519.0,NaN
1,2014-01-09,ANDHRA PRADESH,12234.0,NaN
2,2014-01-10,ANDHRA PRADESH,13349.0,NaN
3,2014-01-11,ANDHRA PRADESH,12437.0,NaN
4,2014-01-12,ANDHRA PRADESH,12040.0,NaN
5,2014-01-13,ANDHRA PRADESH,12152.0,NaN
6,2014-01-14,ANDHRA PRADESH,12008.0,NaN
7,2014-01-15,ANDHRA PRADESH,11159.0,NaN
8,2014-01-16,ANDHRA PRADESH,11158.0,NaN
9,2014-01-17,ANDHRA PRADESH,11285.0,NaN


**Inference:** `pd.concat` vertically combines continuous observations, while `pd.merge` acts like a SQL relational join to enrich the primary dataset with geographic region data.


## Post-Lab Questions & Solutions

### Q1. Explain the concept of "Broadcasting" in NumPy with the help of a suitable example.

NumPy **Broadcasting** describes how arithmetic operations are handled between arrays of different shapes. The smaller array is logically "broadcast" (stretched) across the larger array without making unneeded memory copies, following two rules:
1. Array dimensions are compared element-wise from right to left (trailing dimensions first).
2. Two dimensions are compatible if they are equal, or if one of them is 1.


In [34]:
import numpy as np

# Example of NumPy Broadcasting
demand_matrix = np.array([[1200, 1500, 1800],
                          [2200, 2500, 2800]])  # Shape: (2, 3)
scale_factors = np.array([1.1, 1.2, 1.3])       # Shape: (3,) -> Broadcasts to (2, 3)

scaled_demand = demand_matrix * scale_factors
print("Broadcasted Result:\n", scaled_demand)


Broadcasted Result:
 [[1320. 1800. 2340.]
 [2420. 3000. 3640.]]


### Q2. Explain the difference between label-based indexing (`df.loc[]`) and integer-based indexing (`df.iloc[]`) in Pandas with the help of a suitable example.

| Indexer | Indexing Mode | Slicing Boundary Behavior | Key Use Case |
| :--- | :--- | :--- | :--- |
| **`df.loc[]`** | Explicit **Labels / Names** (Index labels, column names) | **Inclusive** of both start and stop labels | Filtering by date labels, state names, or boolean conditions |
| **`df.iloc[]`** | Implicit **Integer Positions** ($0$ to $n-1$) | **Exclusive** of the stop integer index (Python standard) | Positional extraction (e.g., first 5 rows, last 3 columns) |


In [35]:
# loc vs iloc Demonstration
sample_df = pd.DataFrame({'temp': [32.5, 41.2, 28.0]}, index=['Day_A', 'Day_B', 'Day_C'])

# Label-based: includes 'Day_B'
print("Using loc (Label-based):")
display(sample_df.loc['Day_A':'Day_B', ['temp']])

# Integer-position: index 2 is excluded (retrieves indices 0 and 1)
print("\nUsing iloc (Integer-based):")
display(sample_df.iloc[0:2, [0]])


Using loc (Label-based):


,temp
Day_A,32.5
Day_B,41.2



Using iloc (Integer-based):


,temp
Day_A,32.5
Day_B,41.2


## Conclusion

Through this experiment, practical data handling was executed using **Pandas** and **NumPy** on real-world Indian weather and power grid time-series data (Data used according to the selected usecase). Missing data handling, conditional compound slicing, multi-level aggregations, feature scaling, and relational transformations were systematically applied, and the mechanics of array broadcasting and index slicing were analyzed.
